### PR TITLE
Added http/3 and port reuse support

### PR DESCRIPTION
**Pull request by J3fftw1**
## Type of Change

- **Tool Source:** Vue, JS
- **Something else:** 

## What issue does this relate to?

N/A

### What should this PR do?

This PR adds support for
* HTTP/3
* QUIC

### What are the acceptance criteria?

I changed the https table before it looked like this
alt text
and now it looks like this 
alt text
and it looks like this with http/3 enabled
alt text

The configs looked like this before
alt text
and it looks like this with http/3 enabled
alt text

I also made changes to the language file, I still need some help on adding locale translations

### CODE REVIEW

**Comment by MattIPv4** [comment]
> I also made changes to the language file, I still need some help on adding locale translations

You can just leave them as todo comments for now, with the English strings, e.g.:

`    myKey: 'My English string', // TODO: translate`

**Comment by J3fftw1** [comment]
> 
> 
> > I also made changes to the language file, I still need some help on adding locale translations
> 
> You can just leave them as todo comments for now, with the English strings, e.g.:
> 
> ` myKey: 'My English string', // TODO: translate`

Okay, will do!

Also, do you have any preferences what the message should be?

**Comment by MattIPv4** [comment]
> Also, do you have any preferences what the message should be?

Not sure quite what you mean by message, but if you copy across the English strings and then just add the `// TODO: translate` comment to each line that needs translating in the future, the CI will pick that up :)

**Comment by J3fftw1** [comment]
> > Also, do you have any preferences what the message should be?
> 
> Not sure quite what you mean by message, but if you copy across the English strings and then just add the `// TODO: translate` comment to each line that needs translating in the future, the CI will pick that up :)

I was asking what tgose strings should be. Sorry for the confusing message!

**Comment by MattIPv4** [comment]
Ah! I think the only string missing is the description for `reuseport`, I think something like `enable reuseport to generate a listening socket per worker`

**Comment by J3fftw1** [comment]
> 
> 
> Ah! I think the only string missing is the description for `reuseport`, I think something like `enable reuseport to generate a listening socket per worker`

Sorry, I just saw you commented. @Walshydev really didnt mention it to me 👀

File: src/nginxconfig/i18n/fr/templates/domain_sections/https.js [hunk 1]
```diff
@@ -32,6 +32,8 @@ export default {
     enableHttp2Connections: `${common.enable} les connexions ${common.http}/2`,
     http3: `${common.http}/3`,
     enableHttp3Connections: `${common.enable} les connexions ${common.http}/3`,
+    portReuse: 'My English string', // TODO: translate
```
Comment (WalshyDev) [inline src/nginxconfig/i18n/fr/templates/domain_sections/https.js:35]: He means have this as the English translation haha. So "Reuseport" not actually "My English string"
Numpty

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: src/nginxconfig/i18n/fr/templates/domain_sections/https.js (hunk 1 above)
Comment (J3fftw1) [reply src/nginxconfig/i18n/fr/templates/domain_sections/https.js:35]: oooooooooooooow

Comment (MattIPv4) [reply src/nginxconfig/i18n/fr/templates/domain_sections/https.js:35]: :lul:

Comment (J3fftw1) [reply src/nginxconfig/i18n/fr/templates/domain_sections/https.js:35]: mean :panspoonsad:

File: src/nginxconfig/templates/domain_sections/https.vue [hunk 2]
```diff
@@ -206,6 +243,14 @@ THE SOFTWARE.
             default: true,
             enabled: true,
         },
+        http3: {
+            default: false,
+            enabled: true,
+        },
+        portReuse: {
+            default: true,
```
Comment (MattIPv4) [inline src/nginxconfig/templates/domain_sections/https.vue:251]:
```suggestion
            default: false,
```
We probably don't want reuseport to be on by default

File: src/nginxconfig/templates/domain_sections/https.vue [hunk 3]
```diff
@@ -285,6 +334,20 @@ THE SOFTWARE.
...
+                    if (data.computed) {
+                        this.$props.data.portReuse.enabled = true;
+                        this.$props.data.portReuse.computed = this.$props.data.portReuse.value;
+                    } else {
+                        this.$props.data.portReuse.enabled = false;
+                        this.$props.data.portReuse.computed = false;
+                    }
+                },
+                deep: true,
+            },
```
Comment (MattIPv4) [inline src/nginxconfig/templates/domain_sections/https.vue:350]: Why is this the case? reuseport is completely separate of http2/3?

File: src/nginxconfig/util/angular_backwards_compatibility.js [hunk 4]
```diff
@@ -78,6 +78,8 @@ const domainMap = {
 
     https:                  ['https', 'https', oldBool],
     http2:                  ['https', 'http2', oldBool],
+    http3:                  ['https', 'http3', oldBool],
+    port_reuse:             ['https', 'portReuse', oldBool],
```
Comment (MattIPv4) [inline src/nginxconfig/util/angular_backwards_compatibility.js:82]:
```suggestion
```
These aren't needed as these options weren't present in the angular version of the tool :)

File: src/nginxconfig/templates/domain_sections/https.vue [hunk 5]
```diff
@@ -62,6 +62,43 @@ THE SOFTWARE.
             </div>
         </div>
 
+        <div v-if="http3Enabled" class="field is-horizontal">
+            <div class="field-label">
+                <label class="label">{{ $t('templates.domainSections.https.http3') }}</label>
```
Comment (MattIPv4) [inline src/nginxconfig/templates/domain_sections/https.vue:67]: As http3 is not a standard module for NGINX from my understanding, we need to add a note here (really in the i18n string and then make this `{{{ ... }}}` to render as raw HTML) that links to how to build NGINX with http3 available: https://github.com/cloudflare/quiche/tree/master/extras/nginx

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

File: src/nginxconfig/templates/domain_sections/https.vue (hunk 2 above)
Comment (MattIPv4) [reply src/nginxconfig/templates/domain_sections/https.vue:251]: However, it might be interesting to look at showing a warning if http3 is enabled without reuseport, but as reuseport is a setting separate from http2/3 usage, I think we'll want it always available as an option but therefor not on by default.

File: src/nginxconfig/generators/conf/website.conf.js [hunk 6]
```diff
@@ -60,13 +60,21 @@ const httpsListen = domain => {
...
     // v6
     if (domain.server.listenIpv6.computed)
         config.push(['listen',
-            `[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
-
+            `[${domain.server.listenIpv6.computed}]:443 ssl`
+            + `${domain.https.http2.computed ? ' http2' : '' }`
+            + `${domain.https.http3.computed ? ' http3' : ''}`
+            + `${domain.https.portReuse.computed ? ' reuseport' : ''}`,
+        ]);
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:77]: Doesn't http3 need to be in a separate listen directive as it uses udp rather than tcp?

File: src/nginxconfig/templates/domain_sections/https.vue (hunk 2 above)
Comment (MattIPv4) [reply src/nginxconfig/templates/domain_sections/https.vue:251]: Nvm, ignore me on this, trying to do this is probably going to result in a more complex UI that'll confuse folks.

File: src/nginxconfig/templates/domain_sections/https.vue (hunk 3 above)
Comment (MattIPv4) [reply src/nginxconfig/templates/domain_sections/https.vue:350]: Nvm, ignore me on this, trying to do this is probably going to result in a more complex UI that'll confuse folks.

Comment (J3fftw1) [reply src/nginxconfig/templates/domain_sections/https.vue:350]: I thought they did but apparently not cant find any docs denying or approving it, so /shrug

File: src/nginxconfig/generators/conf/website.conf.js (hunk 6 above)
Comment (J3fftw1) [reply src/nginxconfig/generators/conf/website.conf.js:77]: yes indeed my bad

Comment (J3fftw1) [reply src/nginxconfig/generators/conf/website.conf.js:77]: Is there a way to add comments to the lines in the config. I would like to add UDP listener / TCP listener

Comment (MattIPv4) [reply src/nginxconfig/generators/conf/website.conf.js:77]: Yeah, you can just push a comment:

https://github.com/digitalocean/nginxconfig.io/blob/eb25fc5eeec4ff103dcaa9c164b9d0d3cf873430/src/nginxconfig/generators/conf/website.conf.js#L47

**Comment by J3fftw1** [comment]
alt text
alt text

The images above is how it looks now!

File: src/nginxconfig/templates/domain_sections/https.vue [hunk 7]
```diff
@@ -62,6 +62,43 @@ THE SOFTWARE.
...
+                                <i slot="extra" class="icon fas fa-check"></i>
+                                {{ $t('templates.domainSections.https.enableReuseOfPort') }}
+                            </PrettyCheck>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </div>
+
+
```
Comment (MattIPv4) [inline src/nginxconfig/templates/domain_sections/https.vue:101]:
```suggestion
```
Rm extra linebreak

File: src/nginxconfig/templates/domain_sections/https.vue [hunk 8]
```diff
@@ -188,6 +225,18 @@ THE SOFTWARE.
                 </div>
             </div>
         </div>
+        <template v-if="$props.data.http3.value === true">
```
Comment (MattIPv4) [inline src/nginxconfig/templates/domain_sections/https.vue:228]:
```suggestion

        <template v-if="$props.data.http3.computed">
```
Linebreak + use the computed value so that the notice doesn't show if portuse is enabled but http3 is disabled + don't need to compare to `true`, as true is truthy.

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
Ty 💙

File: src/nginxconfig/templates/domain_sections/https.vue [hunk 9]
```diff
@@ -188,6 +225,18 @@ THE SOFTWARE.
...
+            <br />
+            <div class="message is-warning">
+                <div class="message-body">
+                    {{ $t('templates.globalSections.https.http3Warning1') }}
+                    <ExternalLink :text="$t('templates.globalSections.https.http3Warning2')"
+                                  link="https://github.com/cloudflare/quiche/tree/master/extras/nginx"
```
[the remark this answers is not in the document]
Comment (J3fftw1) [reply src/nginxconfig/templates/domain_sections/https.vue:234]: I wasn't sure about wording anyways. Not my strong side!

File: src/nginxconfig/generators/conf/website.conf.js [hunk 10]
```diff
@@ -60,13 +60,28 @@ const httpsListen = domain => {
     const config = [];
 
     // HTTPS
-    config.push(['listen', `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
+    config.push(['listen', 
+        `${domain.server.listenIpv4.computed === '*' ? '' :`${domain.server.listenIpv4.computed}:`}443 ssl${domain.https.http2.computed ? ' http2' :''}`,
+        ]);
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:65]:
```suggestion
    config.push(['listen', `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
```
Please revert this change so there is a clean diff

File: src/nginxconfig/generators/conf/website.conf.js [hunk 11]
```diff
@@ -60,13 +60,28 @@ const httpsListen = domain => {
...
+            `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 http3` 
+            + `${domain.https.portReuse.computed ? ' reuseport' : ''}`,
+        ]);
+    
     // v6
     if (domain.server.listenIpv6.computed)
         config.push(['listen',
-            `[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
+            `[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : '' }`,
+        ]);
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:78]:
```suggestion
            `[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
```
Please revert this change so there is a clean diff

File: src/nginxconfig/i18n/pt-br/templates/global_sections/https.js [hunk 12]
```diff
@@ -47,4 +47,9 @@ export default {
     ipv4Only: `${ipv4} apenas`,
     ipv6Only: `${ipv6} apenas`,
     ipv4AndIpv6: `${ipv4} & ${ipv6}`,
-};
+    http3Warning1: 'HTTP/3 isn\'t a standard NGINX module, check the ', // TODO: translate 
+    http3Warning2: 'NGINX QUIC readme ', // TODO: translate 
+    http3Warning3: ' or the ', // TODO: translate 
+    http3Warning4: 'Cloudflare quiche project ', // TODO: translate 
+    http3Warning5: ' for how to build NGINX with HTTP/3!', // TODO: translate 
+};
```
Comment (MattIPv4) [inline src/nginxconfig/i18n/pt-br/templates/global_sections/https.js:55]:
```suggestion
};

```
Newline eof

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
Feature lgtm, just some code cleanliness :)

File: src/nginxconfig/generators/conf/website.conf.js [hunk 13]
```diff
@@ -60,13 +60,24 @@ const httpsListen = domain => {
     const config = [];
 
     // HTTPS
-    config.push(['listen', `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
+    config.push(['listen', `${domain.server.listenIpv4.computed === '*' ? '' :`${domain.server.listenIpv4.computed}:`}443 ssl${domain.https.http2.computed ? ' http2' :''}`]);
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:63]: You still have a diff here when nothing should be changing with this line, please revert :)

File: src/nginxconfig/generators/conf/website.conf.js [hunk 14]
```diff
@@ -60,13 +60,24 @@ const httpsListen = domain => {
...
+        config.push(['listen',
+            `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 http3` 
+            + `${domain.https.portReuse.computed ? ' reuseport' : ''}`,
+        ]);
+    
     // v6
     if (domain.server.listenIpv6.computed)
-        config.push(['listen',
-            `[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
+        config.push(['listen',`[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : '' }`]);
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:74]: Ditto

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

File: src/nginxconfig/generators/conf/website.conf.js (hunk 13 above)
Comment (J3fftw1) [reply src/nginxconfig/generators/conf/website.conf.js:63]: Whoops,  sorry for that!

File: src/nginxconfig/generators/conf/website.conf.js [hunk 15]
```diff
@@ -61,12 +61,24 @@ const httpsListen = domain => {
 
     // HTTPS
     config.push(['listen', `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
-
+
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:64]:
```suggestion

```
You've got some trailing whitespace here now :P

File: src/nginxconfig/generators/conf/website.conf.js [hunk 16]
```diff
@@ -61,12 +61,24 @@ const httpsListen = domain => {
...
+            `${domain.server.listenIpv4.computed === '*' ? '' : `${domain.server.listenIpv4.computed}:`}443 http3` 
+            + `${domain.https.portReuse.computed ? ' reuseport' : ''}`,
+        ]);
+    
     // v6
     if (domain.server.listenIpv6.computed)
         config.push(['listen',
             `[${domain.server.listenIpv6.computed}]:443 ssl${domain.https.http2.computed ? ' http2' : ''}`]);
-
+
```
Comment (MattIPv4) [inline src/nginxconfig/generators/conf/website.conf.js:76]:
```suggestion

```
Ditto

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

**Comment by WalshyDev** [comment]
I love this PR honestly :lul:

Jeff ❤️

**Comment by J3fftw1** [comment]
> 
> 
> I love this PR honestly :lul:
> 
> Jeff ❤️

Now this is bullying
/unfriend @WalshyDev

**Review by MattIPv4 — APPROVED** [review]
Lgtm 🎉

**Comment by J3fftw1** [comment]
> 
> 
> Lgtm 🎉

Woohoo

**Comment by MattIPv4** [comment]
Hey, @J3fftw1 - Thanks a ton for this sweet PR! 😄

Would you please shoot me an email when you get a chance?
mcowley at digitalocean dot com 🎉

**Comment by J3fftw1** [comment]
> 
> 
> Hey, @J3fftw1 - Thanks a ton for this sweet PR! 😄
> 
> Would you please shoot me an email when you get a chance?
> mcowley at digitalocean dot com 🎉

Hey @MattIPv4 No problem was fun working on it even though all the tiny mistakes I made.
Was also my first time actually working on something JS was fun learning aswell. I will shoot you an email right now!
